tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,577 @@
+      if (error != null) {
+        return error;
+      }
+    }
+    return null;
+  }
+
+  private String doCommitSecondaryKeySingleBatch(
+      BackOffer backOffer, BatchKeys batchKeys, long commitTs) {
+    List<byte[]> keysCommit = batchKeys.getKeys();
+    byte[][] keys = new byte[keysCommit.size()][];
+    keysCommit.toArray(keys);
+    // send rpc request to tikv server
+    long regionId = batchKeys.getRegioId();
+    ClientRPCResult commitResult =
+        this.kvClient.commit(backOffer, keys, this.startTs, commitTs, regionId);
+    if (!commitResult.isSuccess() && !commitResult.isRetry()) {
+      String error =
+          String.format("Txn commit secondary key error, regionId=%s", batchKeys.getRegioId());
+      LOG.warn(error);
+      return error;
+    }
+    if (!commitResult.isSuccess() && commitResult.isRetry()) {
+      try {
+        backOffer.doBackOff(
+            BackOffFunction.BackOffFuncType.BoRegionMiss,
+            new GrpcException(
+                String.format(
+                    "Txn commit secondary key failed, regionId=%s", batchKeys.getRegioId())));
+        // re-split keys and commit again.
+        int size = batchKeys.getKeys().size();
+        byte[][] keyBytes = new byte[size][];
+        int i = 0;
+        for (byte[] k : batchKeys.getKeys()) {
+          keyBytes[i] = k;
+          i++;
+        }
+        String error = this.doCommitSecondaryKeys(backOffer, keyBytes, size, commitTs);
+        if (error != null) {
+          LOG.warn(error);
+          return error;
+        }
+      } catch (GrpcException e) {
+        String error =
+            String.format(
+                "Txn commit secondary key error, re-split commit failed, regionId=%s",
+                batchKeys.getKeys());
+        LOG.warn(error);
+        return error;
+      }
+    }
+    if (!commitResult.isSuccess()) {
+      String error =
+          String.format(
+              "Txn commit secondary key error, regionId=%s, detail=%s",
+              batchKeys.getKeys(), commitResult.getError());
+      LOG.warn(error);
+      return error;
+    }
+
+    LOG.info("commit " + batchKeys.getKeys().size() + "rows successfully");
+    return null;
+  }
+
+  private GroupKeyResult groupKeysByRegion(BackOffer backOffer, byte[][] keys, int size) {
+    Map<Long, List<byte[]>> groups = new HashMap<>();
+    long first = 0;
+    int index = 0;
+    String error = null;
+    try {
+      for (; index < size; index++) {
+        byte[] key = keys[index];
+        TiRegion tiRegion = this.regionManager.getRegionByKey(ByteString.copyFrom(key));
+        if (tiRegion != null) {
+          Long regionId = tiRegion.getId();
+          if (index == 0) {
+            first = regionId;
+          }
+          List<byte[]> groupItem = groups.computeIfAbsent(regionId, e -> new LinkedList<>());
+          groupItem.add(key);
+          groups.put(tiRegion.getId(), groupItem);
```

Comment:
```suggestion
          groups.computeIfAbsent(regionId, e -> new LinkedList<>()).add(key);
```
your code put the groupItem into groups twice.